src/main/java/com/aws/greengrass/shadowmanager/sync/IotDataPlaneClientFactory.java

```diff
@@ -55,20 +60,22 @@ public class IotDataPlaneClientFactory {
     private static final Set<Class<? extends Exception>> retryableIoTExceptions = new HashSet<>(
             Arrays.asList(ThrottlingException.class, InternalException.class, InternalFailureException.class,
                     LimitExceededException.class));
+    private final DeviceConfiguration deviceConfiguration;
 
     /**
      * Constructor for IotDataPlaneClientFactory to maintain IoT Data plane client.
      *
      * @param deviceConfiguration Device configuration class.
      */
     @Inject
+    @SuppressWarnings({"PMD.NullAssignment"})
     public IotDataPlaneClientFactory(DeviceConfiguration deviceConfiguration) {
-        configureClient(deviceConfiguration);
+        this.deviceConfiguration = deviceConfiguration;
         deviceConfiguration.onAnyChange((what, node) -> {
             if (validString(node, DEVICE_PARAM_AWS_REGION) || validPath(node, DEVICE_PARAM_ROOT_CA_PATH) || validPath(
                     node, DEVICE_PARAM_CERTIFICATE_FILE_PATH) || validPath(node, DEVICE_PARAM_PRIVATE_KEY_PATH)
                     || validString(node, DEVICE_PARAM_IOT_DATA_ENDPOINT)) {
-                configureClient(deviceConfiguration);
+                iotDataPlaneClient = null;
```

Comment:
now this isn't threadsafe, so you need to use an atomic or synchronize here.

Reply:
updated the PR with atomic client